### ISSUE TITLE
US Holidays Error: dictionary changed during iteration when passing years to holidays.US constructor

### ISSUE THREAD

**Issue by extrospective** [dr-prodigy/python-holidays#569]
Attempting to use code to get US holidays for specific years.  This code works for some countries but not US (nor Canada)

```
import holidays
import numpy as np
print(holidays.__version__)
years = np.array([2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021])
x = holidays.US(years=years)
```

This code works when I try different countries: FR, GB, RU, DE, ES, CZ, AU, JP, CN, IN

This syntax **fails** to return results for the countries: US, CA

This syntax for passing numpy arrays I found in a ticket with a CZ example.  (https://github.com/dr-prodigy/python-holidays/pull/528).

I am running on Databricks runtime 10.0.  The holidays version is 0.11.3.1 as you see below.

The output I got was for US (or CA) is:
```
0.11.3.1
Out[41]: 
RuntimeError: dictionary changed size during iteration
```

Example output when I use another country, such as AU is:
```
0.11.3.1
Out[42]: {datetime.date(2016, 1, 1): "New Year's Day",
 datetime.date(2016, 1, 26): 'Australia Day',
 datetime.date(2016, 3, 25): 'Good Friday',
 datetime.date(2016, 3, 28): 'Easter Monday',
 datetime.date(2016, 4, 25): 'Anzac Day',
 datetime.date(2016, 12, 25): 'Christmas Day',
 datetime.date(2016, 12, 27): 'Christmas Day (Observed)',
 datetime.date(2016, 12, 26): 'Boxing Day',
 datetime.date(2017, 1, 1): "New Year's Day",
 datetime.date(2017, 1, 2): "New Year's Day (Observed)",
...
```

**Comment by dr-prodigy** [comment]
Hi @extrospective
thank you for your interest in python-holidays.

I've made a test on your code, but I can't seem to replicate your issue, neither with US, nor with CA.
Here's my result:

```
>>> import holidays
>>> import numpy as np
>>> print(holidays.__version__)
0.11.3.1
>>> years = np.array([2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021])
>>> x = holidays.US(years=years)
>>> x
{datetime.date(2016, 1, 1): "New Year's Day", datetime.date(2016, 1, 18): 'Martin Luther King Jr. Day', datetime.date(2016, 2, 15): "Washington's Birthday", datetime.date(2016, 5, 30): 'Memorial Day', datetime.date(2016, 7, 4): 'Independence Day', datetime.date(2016, 9, 5): 'Labor Day', datetime.date(2016, 10, 10): 'Columbus Day', datetime.date(2016, 11, 11): 'Veterans Day', datetime.date(2016, 11, 24): 'Thanksgiving', datetime.date(2016, 12, 25): 'Christmas Day', datetime.date(2016, 12, 26): 'Christmas Day (Observed)', datetime.date(2017, 1, 1): "New Year's Day", datetime.date(2017, 1, 2): "New Year's Day (Observed)", datetime.date(2017, 1, 16): 'Martin Luther King Jr. Day', datetime.date(2017, 2, 20): "Washington's Birthday", datetime.date(2017, 5, 29): 'Memorial Day', datetime.date(2017, 7, 4): 'Independence Day', datetime.date(2017, 9, 4): 'Labor Day', datetime.date(2017, 10, 9): 'Columbus Day', datetime.date(2017, 11, 11): 'Veterans Day', datetime.date(2017, 11, 10): 'Veterans Day (Observed)', datetime.date(2017, 11, 23): 'Thanksgiving', datetime.date(2017, 12, 25): 'Christmas Day', datetime.date(2018, 1, 1): "New Year's Day", datetime.date(2018, 1, 15): 'Martin Luther King Jr. Day', datetime.date(2018, 2, 19): "Washington's Birthday", datetime.date(2018, 5, 28): 'Memorial Day', datetime.date(2018, 7, 4): 'Independence Day', datetime.date(2018, 9, 3): 'Labor Day', datetime.date(2018, 10, 8): 'Columbus Day', datetime.date(2018, 11, 11): 'Veterans Day', datetime.date(2018, 11, 12): 'Veterans Day (Observed)', datetime.date(2018, 11, 22): 'Thanksgiving', datetime.date(2018, 12, 25): 'Christmas Day', datetime.date(2019, 1, 1): "New Year's Day", datetime.date(2019, 1, 21): 'Martin Luther King Jr. Day', datetime.date(2019, 2, 18): "Washington's Birthday", datetime.date(2019, 5, 27): 'Memorial Day', datetime.date(2019, 7, 4): 'Independence Day', datetime.date(2019, 9, 2): 'Labor Day', datetime.date(2019, 10, 14): 'Columbus Day', datetime.date(2019, 11, 11): 'Veterans Day', datetime.date(2019, 11, 28): 'Thanksgiving', datetime.date(2019, 12, 25): 'Christmas Day', datetime.date(2020, 1, 1): "New Year's Day", datetime.date(2020, 1, 20): 'Martin Luther King Jr. Day', datetime.date(2020, 2, 17): "Washington's Birthday", datetime.date(2020, 5, 25): 'Memorial Day', datetime.date(2020, 7, 4): 'Independence Day', datetime.date(2020, 7, 3): 'Independence Day (Observed)', datetime.date(2020, 9, 7): 'Labor Day', datetime.date(2020, 10, 12): 'Columbus Day', datetime.date(2020, 11, 11): 'Veterans Day', datetime.date(2020, 11, 26): 'Thanksgiving', datetime.date(2020, 12, 25): 'Christmas Day', datetime.date(2021, 1, 1): "New Year's Day", datetime.date(2021, 12, 31): "New Year's Day (Observed)", datetime.date(2021, 1, 18): 'Martin Luther King Jr. Day', datetime.date(2021, 2, 15): "Washington's Birthday", datetime.date(2021, 5, 31): 'Memorial Day', datetime.date(2021, 6, 19): 'Juneteenth National Independence Day', datetime.date(2021, 7, 4): 'Independence Day', datetime.date(2021, 7, 5): 'Independence Day (Observed)', datetime.date(2021, 9, 6): 'Labor Day', datetime.date(2021, 10, 11): 'Columbus Day', datetime.date(2021, 11, 11): 'Veterans Day', datetime.date(2021, 11, 25): 'Thanksgiving', datetime.date(2021, 12, 25): 'Christmas Day', datetime.date(2021, 12, 24): 'Christmas Day (Observed)', datetime.date(2011, 1, 1): "New Year's Day", datetime.date(2010, 12, 31): "New Year's Day (Observed)", datetime.date(2011, 1, 17): 'Martin Luther King Jr. Day', datetime.date(2011, 2, 21): "Washington's Birthday", datetime.date(2011, 5, 30): 'Memorial Day', datetime.date(2011, 7, 4): 'Independence Day', datetime.date(2011, 9, 5): 'Labor Day', datetime.date(2011, 10, 10): 'Columbus Day', datetime.date(2011, 11, 11): 'Veterans Day', datetime.date(2011, 11, 24): 'Thanksgiving', datetime.date(2011, 12, 25): 'Christmas Day', datetime.date(2011, 12, 26): 'Christmas Day (Observed)', datetime.date(2012, 1, 1): "New Year's Day", datetime.date(2012, 1, 2): "New Year's Day (Observed)", datetime.date(2012, 1, 16): 'Martin Luther King Jr. Day', datetime.date(2012, 2, 20): "Washington's Birthday", datetime.date(2012, 5, 28): 'Memorial Day', datetime.date(2012, 7, 4): 'Independence Day', datetime.date(2012, 9, 3): 'Labor Day', datetime.date(2012, 10, 8): 'Columbus Day', datetime.date(2012, 11, 11): 'Veterans Day', datetime.date(2012, 11, 12): 'Veterans Day (Observed)', datetime.date(2012, 11, 22): 'Thanksgiving', datetime.date(2012, 12, 25): 'Christmas Day', datetime.date(2013, 1, 1): "New Year's Day", datetime.date(2013, 1, 21): 'Martin Luther King Jr. Day', datetime.date(2013, 2, 18): "Washington's Birthday", datetime.date(2013, 5, 27): 'Memorial Day', datetime.date(2013, 7, 4): 'Independence Day', datetime.date(2013, 9, 2): 'Labor Day', datetime.date(2013, 10, 14): 'Columbus Day', datetime.date(2013, 11, 11): 'Veterans Day', datetime.date(2013, 11, 28): 'Thanksgiving', datetime.date(2013, 12, 25): 'Christmas Day', datetime.date(2014, 1, 1): "New Year's Day", datetime.date(2014, 1, 20): 'Martin Luther King Jr. Day', datetime.date(2014, 2, 17): "Washington's Birthday", datetime.date(2014, 5, 26): 'Memorial Day', datetime.date(2014, 7, 4): 'Independence Day', datetime.date(2014, 9, 1): 'Labor Day', datetime.date(2014, 10, 13): 'Columbus Day', datetime.date(2014, 11, 11): 'Veterans Day', datetime.date(2014, 11, 27): 'Thanksgiving', datetime.date(2014, 12, 25): 'Christmas Day', datetime.date(2015, 1, 1): "New Year's Day", datetime.date(2015, 1, 19): 'Martin Luther King Jr. Day', datetime.date(2015, 2, 16): "Washington's Birthday", datetime.date(2015, 5, 25): 'Memorial Day', datetime.date(2015, 7, 4): 'Independence Day', datetime.date(2015, 7, 3): 'Independence Day (Observed)', datetime.date(2015, 9, 7): 'Labor Day', datetime.date(2015, 10, 12): 'Columbus Day', datetime.date(2015, 11, 11): 'Veterans Day', datetime.date(2015, 11, 26): 'Thanksgiving', datetime.date(2015, 12, 25): 'Christmas Day'}
```

Same goes with the latest beta version:

```
>>> import holidays
>>> import numpy as np
>>> print(holidays.__version__)
0.11.4
>>> years = np.array([2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021])
>>> x = holidays.US(years=years)
>>> x
{datetime.date(2016, 1, 1): "New Year's Day", datetime.date(2016, 1, 18): 'Martin Luther King Jr. Day', datetime.date(2016, 2, 15): "Washington's Birthday", datetime.date(2016, 5, 30): 'Memorial Day', datetime.date(2016, 7, 4): 'Independence Day', datetime.date(2016, 9, 5): 'Labor Day', datetime.date(2016, 10, 10): 'Columbus Day', datetime.date(2016, 11, 11): 'Veterans Day', datetime.date(2016, 11, 24): 'Thanksgiving', datetime.date(2016, 12, 25): 'Christmas Day', datetime.date(2016, 12, 26): 'Christmas Day (Observed)', datetime.date(2017, 1, 1): "New Year's Day", datetime.date(2017, 1, 2): "New Year's Day (Observed)", [...]
```

I then assume that behaviour has to do with Databricks runtime, which I really don't know.
If you could investigate further / figure out what's wrong, I would really appreciate it.. Thx a lot! 👍

**Comment by lior-lemonade** [comment]
I got the same error (when using FBProphet). 

```
from holidays.countries import US
year_list=[2022]
holidays = US(prov=None, state=None, years=year_list)
[(date, holidays.get_list(date)) for date in holidays]
```

**Comment by mborsetti** [comment]
@dr-prodigy I was able to replicate the issue:
```
>>> from holidays.countries import US
>>>holidays = US(prov=None, state=None, years=[2022])
>>>len(holidays)
14
>>>[(date, holidays.get_list(date)) for date in holidays]
Traceback (most recent call last):
  File "C:\Program Files\Python\Python39\lib\site-packages\IPython\core\interactiveshell.py", line 3457, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-75567ae737d3>", line 1, in <module>
    [(date, holidays.get_list(date)) for date in holidays]
  File "<ipython-input-5-75567ae737d3>", line 1, in <listcomp>
    [(date, holidays.get_list(date)) for date in holidays]
RuntimeError: dictionary changed size during iteration
>>>len(holidays)
28
```

I will try to find out why the dictionary expands to include 2021 when invoked in the list-building; my hunch is that it's somehow triggered by one of the `holidays.get_list(date)` calls.  

@dr-prodigy  Do you remember why there's a `get_list()` method when the holiday's name is already the value of the dictionary entry? And why is the value a string...what happens when multiple holidays fall on the same day? Confused.

**Comment by dr-prodigy** [comment]
Hi @mborsetti
thank you for your contribution.. I think I need some more explanation here!

Although there could probably be sth to be checked in the year boundaries logic (ie: 2021 holidays in the list), if I caught it right, your code is "acting funny" (probably on purpose), in that it forces exactly a holiday list recalculation for each date in the previously initialized list itself.. which is why the dictionary changes!
Even it is not occurring on any country, IMHO it is therefore more than expectable to see that error.
I also don't find a match with @extrospective code, which is instead cleanly initializing the US holidays object providing an array of years.

If I'm mistaking / losing something on the way, please explain better.. thank you very much! :)

**Comment by mborsetti** [comment]
@dr-prodigy The code is simply a restatement of the funny code of @lior-lemonade.

Turns out that we were doing this in `UnitedStates(HolidayBase)` (and a few others):
```
            elif self.observed and date(year, JAN, 1).weekday() == SAT:
                # Add Dec 31st from the previous year without triggering
                # the entire year to be added
                expand = self.expand
                self.expand = False
                self[date(year, JAN, 1) + rd(days=-1)] = name + " (Observed)"
                self.expand = expand
```
So if you were asking for a year where 1 January was a Saturday (2021, for example), you'd also be getting a holiday in 2020(!).  The code from @lior-lemonade would then call ``holidays.get_list(date)`` on a date that was in 2020, which would trigger the expansion into 2020 that would change the size of the dictionary.

Turns out that code is not at all required, because we also had this immediately following, which put the observed New Year's day into the correct year:
```
            # The next year's observed New Year's Day can be in this year
            # when it falls on a Friday (Jan 1st is a Saturday)
            if self.observed and date(year, DEC, 31).weekday() == FRI:
                self[date(year, DEC, 31)] = name + " (Observed)"
```

I have worked a PR that addresses this issue and in general ensures that only holidays for the year requested are returned (no expansion into other years)!

**Comment by extrospective** [comment]
I just tried in my conda environment (Windows; not databricks) and got the same error:

Code:
```
import holidays
import numpy as np
print(holidays.__version__)
years = np.array([2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021])
x = holidays.US(years=years)
x
```

Code says I am using version 0.11.3.1

I get an error stack:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
d:\Anaconda3\envs\py39\lib\site-packages\IPython\core\formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

d:\Anaconda3\envs\py39\lib\site-packages\IPython\lib\pretty.py in pretty(self, obj)
    375                 if cls in self.type_pprinters:
    376                     # printer registered in self.type_pprinters
--> 377                     return self.type_pprinters[cls](obj, self, cycle)
    378                 else:
    379                     # deferred printer

d:\Anaconda3\envs\py39\lib\site-packages\IPython\lib\pretty.py in inner(obj, p, cycle)
    599         p.begin_group(step, start)
    600         keys = obj.keys()
--> 601         for idx, key in p._enumerate(keys):
    602             if idx:
    603                 p.text(',')

d:\Anaconda3\envs\py39\lib\site-packages\IPython\lib\pretty.py in _enumerate(self, seq)
    278     def _enumerate(self, seq):
    279         """like enumerate, but with an upper limit on the number of items"""
--> 280         for idx, x in enumerate(seq):
    281             if self.max_seq_length and idx >= self.max_seq_length:
    282                 self.text(',')

RuntimeError: dictionary changed size during iteration
```

**Comment by extrospective** [comment]
jupyter gave me a nicer call stack perhaps than databricks; could have saved some time here.  but glad it seems like another bug was surfaced too.

**Comment by mborsetti** [comment]
@extrospective Still cannot reproduce, at least with the currently published version (0.12):

**Comment by mborsetti** [comment]
@extrospective Still cannot reproduce, at least with the currently published version (0.12):

<img width="770" alt="image" src="https://user-images.githubusercontent.com/26698111/148332984-1e2eafe8-57f2-4d5c-a530-650b808d582f.png">

**Comment by extrospective** [comment]
Just ran in Jupyter with 0.12.
It works correctly.  Issue appears to be in 0.11.3.1 but not in 0.12.

Evidence of success pasted here:
<img width="1587" alt="image" src="https://user-images.githubusercontent.com/11449919/148381433-a189c5f0-4909-48cc-8e93-45a4e9fa3787.png">

**Comment by extrospective** [comment]
Perhaps this fix should be added to release notes for 0.12.

**Comment by dr-prodigy** [comment]
Sorry @extrospective your last suggestion went unnoticed while working on v.0.12.
I reviewed CHANGES file accordingly and released it on beta branch (master will be updated upon next official release).
Thx for your support, cheers!
